### DISCUSSION TITLE
How to Navigate to different page in avalonia UI using Community Toolkit MVVM package

### DISCUSSION THREAD

**Discussion by buminduyasith** [AvaloniaUI/Avalonia#9988] [Q&A]
I have created a project using Avalonia UI MVVM C# project. and I am very new to this. I am using the Community Toolkit MVVM package, not the Reactive UI package. There are two pages, "MainWindow" and "SettingWindow". There is a button on the "MainWindow" page. Once the user clicks that button, user should be navigated to the "Settings" page. How can I accomplish this using the Community Toolkit MVVM package in Avalonia UI?
MainWindow codebehind
```
public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            DataContext= new MainWindowViewModel();
        }
    }
```
MainWindow viewModel Code

```
   [INotifyPropertyChanged]
    public partial class MainWindowViewModel
    {
        [ObservableProperty]
        public string greeting = "Welcome to Avalonia!";

        [RelayCommand]
        public void GotoSettings()
        {
          //navigate to setting window
        }
    }
```

SettingWindow codebehind

```
 public partial class SettingWindow : Window
        {
            public SettingWindow()
            {
                InitializeComponent();
                DataContext = new SettingWindowViewModel();
            }
        }
```

SettingWindow ViewModel code

```
[INotifyPropertyChanged]
public partial class SettingWindowViewModel
{
    public string greeting => "setting page";
}

```

**Comment by timunie** [comment]
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample

Maybe this sample helps you.

**Reply by buminduyasith** [reply]
Thank you for your reply, but that documentation also used Reactive UI.  I want to know how to navigate between pages using Community tool kit mvvm

**Reply by timunie** [reply]
It should not make much a difference tbh. The concept should be similar. You just use [RelayCommand] instead. DataTemplates etc are not affacted.

**Reply by buminduyasith** [reply]
Yeah, this worked for me. but the issue is it's challenging to do the dependency injection.

 ```
 public string Greeting => "Welcome to Avalonia!";
        ICalCulationService calCulationService = new CalCulationService();
        List<PageViewModelBase> Pages = new List<PageViewModelBase>();

        public MainWindowViewModel()
        {
            var firstPageViewModel = new FirstPageViewModel(calCulationService);
            var secondPageViewModel = new SecondPageViewModel();
            Pages.Add(firstPageViewModel);
            Pages.Add(secondPageViewModel);
            CurrentPage = Pages[0];
          
        }

```
do you know any different method to do this

**Reply by sandreas** [reply]
> do you know any different method to do this

Usually, the Dependency Injection happens BEFORE and the instance creation of the view model is done by the IoC container, a Locator or a Factory (depending on your preferences). A very simple and naive approach could be this (below), while more advanced techniques would be (see links)
- https://stackoverflow.com/questions/71119254/avaloniaui-what-is-the-proper-way-to-inject-viewmodels-into-views-using-compos
- https://dev.to/ingvarx/avaloniaui-dependency-injection-4aka
```c#
// first register the view model as "service" (e.g. services.AddScoped<MainWindowViewModel>))
namespace RoutingExample
{
    public class MainWindowViewModel : ReactiveObject, IScreen
    {
        public MainWindowViewModel()
        {
             public string Greeting => "Welcome to Avalonia!";

             ICalCulationService calCulationService;

             List<PageViewModelBase> Pages = new List<PageViewModelBase>();
      
             public MainWindowViewModel(ICalCulationService cal, FirstPageViewModel first, SecondPageViewModel second)
             {
                 calCulationService = cal;
                 Pages.Add(first);
                 Pages.Add(second);
                 CurrentPage = first;
             }
        }
    }
}

```

**Reply by puppy1980** [reply]
If you want to navigate like maui, you can use community mvvm like this:

1. creat static class NavigateHelper<T>(Dictionry<string,object> dic=null,string container=null) where T:Control
{
desktop.mainwindow.content=App.Current.Services.GetService<T>();//use inject loc
}

2.creat   [AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
    public class QueryAttribute : Attribute
    {
        public string QueryName { get; set; }
        public string PropertyName { get; set; }

        public QueryAttribute(string queryName, string propertyName)
        {
            QueryName = queryName;
            PropertyName = propertyName;
        }
    }
if dic!=null  targetviewmodel set that ,then use reflection set it 
3. If you want to navigate within page element, set a contentcontrol and give the name to the container

**Reply by sandreas** [reply]
@puppy1980 
I, personally, dislike the `MAUI` Routing approach. Maybe there is something I don't understand, but I find it WAY too complex. 
For avalonia, I've written a small library, that worked pretty well for my use cases: 

https://github.com/sandreas/Avalonia.SimpleRouter

Feedback welcome.

**Reply by Yveske1966** [reply]
@puppy1980 Could you please help me with your code in point 1 ?
I've pasted the code at the end of the file App.axaml.cs but I get a lot of errors.
Is this the correct place ?
Does it need additional using statements ?

Could you also elaborate on what is meant with "if dic!=null targetviewmodel set that, then use reflection set it"

I'm new to C# and made my first steps with Maui, which I very much liked until I hit a wall, and would like to reuse as much code as I can.
ReactUI is just too verbose and unreadable for me.

I guess newbies like me would benefit from a small demo program with a couple views with dependency injection and a couple of buttons to switch between them.

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#%EF%B8%8F-routing-navigation-samples

**Reply by Yveske1966** [reply]
@timunie I've seen this but what I'm looking for is a **complete** running demo program using Avalonia with community toolkit Maui-style with **dependency injection** and **parameter passing**.

**Reply by timunie** [reply]
check also Awsome-Avalonia.

**Comment by kyuranger** [comment]
I'm not sure if this is what you want.

Firstly, set both pages to UserControl.

And here is XAML:
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication2.MainWindow"
        Title="AvaloniaApplication2">
	<Grid RowDefinitions="auto,*">
		<Button Name="B">Change</Button>
		<ContentControl Content="{Binding Content}"></ContentControl>
	</Grid>
</Window>
```

Here is Code-behind:
```
using Avalonia.Controls;
using Avalonia.Media;
using System;

namespace AvaloniaApplication2
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            B.Click=B_Click;
            this.DataContext = new MainWindowViewModel();            
        }


        private void B_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            if (this.DataContext is MainWindowViewModel i)
            {
                i.Content = new SettingControl();
            }
        }
    }
}
```

Here is ViewModel:
```
using CommunityToolkit.Mvvm.ComponentModel;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AvaloniaApplication2
{
    public partial class MainWindowViewModel:ObservableObject
    {
        [ObservableProperty]
        object? _Content = new MainWindowControl();
    }
}

```

Hope this helps

**Reply by Yveske1966** [reply]
@kyuranger Thank you for answer but the XAML results in Invalid Markup and building results in the following errors:
Error CS0079 The event 'Button.Click' can only appear on the left hand side of += or -=
Error CS0246 The type or namespace name 'MainWindowControl' could not be found (are you missing a using directive or an assembly reference?)
Error Unable to find a setter that allows multiple assignments to the property Content of type

What I'm looking for is a **complete** running demo program using Avalonia with community toolkit Maui-style with **dependency injection** and **parameter passing**. Right now it is very frustrating to get incomplete non-working small pieces of different puzzles, which don't fit together.

**Reply by sandreas** [reply]
@Yveske1966 
Did you checkout https://github.com/sandreas/Avalonia.SimpleRouter ?

The Readme shows a complete working example code, if you wanna see it in Action, you could take a look at my [ToneAudioPlayer Project](https://github.com/sandreas/ToneAudioPlayer/blob/main/ToneAudioPlayer/ViewModels/MainViewModel.cs)...

**Reply by Yveske1966** [reply]
Thanks, will have a look into it and from what I see it looks promising !
Will definitly have a look at your project !

**Reply by sandreas** [reply]
> Thanks, will have a look into it and from what I see it looks promising ! Will definitly have a look at your project !

@Yveske1966 

Yeah its not that impressive :-) It's a set of not more than 2 classes with no dependencies. Should be easy to copy them over into your project, if you don't want additional dependencies, but there is a [nuget package](https://www.nuget.org/packages/Sandreas.Avalonia.SimpleRouter) depending on `Microsoft.Extensions.DependencyInjection.Abstractions` to have a bit more convenience.

The most fun part with CommunityToolkit code generation is this:

```c#
public partial class MainViewModel : ViewModelBase
{
       
    [ObservableProperty]
    private ViewModelBase _content = default!;

    public MainViewModel(HistoryRouter<ViewModelBase> router)
    {
        // register route changed event to set content to viewModel, whenever 
        // a route changes
        router.CurrentViewModelChanged += viewModel => Content = viewModel;
        
        // change to HomeView and store view model for later
        var homeViewModel = router.GoTo<HomeViewModel>();
        
       // set "parameters" in homeviewmodel without having to use string parameters or something strange...
       homeViewModel.Greeting = "Hello";
    }
}
```

So you basically won't really need a complex parameters implementation. You can just set Properties directly within the ViewModel and thanks to generics you have full type support including code completition. Maybe there are caveats or race conditions but for me it works.

**Comment by Yveske1966** [comment]
Thanks again, converting my project from Maui to Avalonia now...  a lot is the same but different.
Still have some work todo, but will continue tomorrow... need a break.

**Reply by Yveske1966** [reply]
Your program works beautifully (I can switch views), but mine doesn't :(
The nuget package SimpleRouter was added to the program.

The following code is called:
`services.AddSingleton<HistoryRouter<ViewModelBase>> (s => new HistoryRouter<ViewModelBase> (t => (ViewModelBase) s.GetRequiredService (t)));`

Result:
`AssemblyQualifiedName = "SelectPicturesAvalonia.ViewModels.SettingsViewModel, SelectPicturesAvalonia, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"`

Afterwards router.HasPrev = false; (true in your code)
ViewLocator is never called in mine, so I guess the view corresponding with the viewmodel isn't called.

What I also find odd is that although all my views are physically in the directory myprogram.Views, within the program only MainView and MainWindow are found on that path, the other ones are found in myprogram.

Is there anything else that I can check ?

**Reply by sandreas** [reply]
You have to 

- Create a `ViewModelBase` extending `ObservableObject` from CommunityToolkit (see https://github.com/sandreas/ToneAudioPlayer/blob/main/ToneAudioPlayer/ViewModels/ViewModelBase.cs#L5)
- Register all ViewModels as a service in `ServiceCollection`, `MainViewModel` as singleton, the others as transient (see https://github.com/sandreas/ToneAudioPlayer/blob/8fa148eaf3910ca075c4d9c6e17e152a8b98d9df/ToneAudioPlayer/App.axaml.cs#L105C22-L105C22)
- Add the ViewLocator to your `App.axaml` (see https://github.com/sandreas/ToneAudioPlayer/blob/8fa148eaf3910ca075c4d9c6e17e152a8b98d9df/ToneAudioPlayer/App.axaml#L9)
- Add a `ContentControl` bound to `Content` (`_content` will generate this property automatically for you) to your `MainView.axaml` (see https://github.com/sandreas/ToneAudioPlayer/blob/8fa148eaf3910ca075c4d9c6e17e152a8b98d9df/ToneAudioPlayer/Views/MainView.axaml#L10)
- Connect your `MainViewModel` with `Content` to the History Router (see https://github.com/sandreas/ToneAudioPlayer/blob/8fa148eaf3910ca075c4d9c6e17e152a8b98d9df/ToneAudioPlayer/ViewModels/MainViewModel.cs#L16)

That should be it.

**Reply by Yveske1966** [reply]
Thanks ! The ContentControl was missing, my bad !

**Reply by thevortexcloud** [reply]
On a side note, if you are interested I have just hacked together a basic router that is path based (think kind of like a website URL). I am not sure how ready it is for people to use but it does work. If you want the code for it, let me know.

It's implemented entirely in .NET standard 2.0 (meaning it will also work with .NET framework) and is not tied to a specific framework. It's not particularly advanced right now but it's fairly flexible. The main advantage this has is you can for the most part, easily use it entirely from within XAML since it's not directly tied to view models like most other routers are. EG this is how I have am using it in my Avalonia test app:

```xaml
            <Button Content="Main" Command="{CompiledBinding GoToPageCommand}" CommandParameter="ViewModels/First" />
            <Button Content="Test" Command="{CompiledBinding GoToPageCommand}" CommandParameter="ViewModels/Second" />
            <Button Content="Back" Command="{CompiledBinding GoBackCommand}" />
```
It can also automatically discovers all the routes too based on an attribute on the class. EG this is one of the view models

```csharp
    [RelativeRoute("First")]
    public class FirstViewModel : ViewModelBase { }
```

Manual registration of routes is also possible.

That said,  the big disadvantage of this is it is path based and therefore not strongly typed. Although that could probably be fixed by binding to some static value on the view model and not using the attributes at all. But I am still messing around with it. 

Implementing some sort of reverse lookup to also work directly with the view model type should also be possible.

**Reply by Yveske1966** [reply]
I'm good for now, pagination works fine now, got my views physically and logically in the Views directory by moving them out and back in.
Still need to translate a lot of code/xaml from Maui to Avalonia, but that should be straightforward (I hope).

I'm finding Avalonia a lot finickier and easier to blow up with lots of strange errors that won't go away.
The only solution I found was excluding a file from the project and adding it back in without further changing anything.
Lost a lot of time (and hair) that way...

Thanks again for your extensive support !

**Reply by thevortexcloud** [reply]
> I'm finding Avalonia a lot finickier and easier to blow up with lots of strange errors that won't go away.

There are apparently some changes in nightly builds to give significantly more verbose error messages. But I have not tried them myself. Specifically you put

```xml
<AvaloniaXamlReportImportance>High</AvaloniaXamlReportImportance>
<AvaloniaXamlVerboseExceptions>true</AvaloniaXamlVerboseExceptions>
```

In your project file.

> Thanks again for your extensive support !

You are welcome!
